# Примеры преобразования размерных величин 
Преобразования размерных величин удобно выполнять с модулем символьных вычислений `python` - `sympy`. 
Преобразования размерностей ключевых величин полезно знать наизусть, хотя всегда можно найти их в таблицах. Значения многих физические константы зашины в модуле `scipy.constants`, иногда это оказывается удобным, при этом автоматически будет учитываться достаточно большое количество знаков после запятой в константах.
Рассмотрим размерности ряда величин широко применяемых в нефтяном инжиниринге.


In [1]:
import sympy as sp
import scipy.constants as const

In [2]:
# в модуле scipy.constants есть значения общепринятых констант -например значение pi
print(const.pi)

3.141592653589793


### Объемный расход $q$

В СИ измеряется в [м$^3$/сек], в практических метрических единицах измеряется в [м$^3$/сут], в американских промысловых единицах измеряется в [bbl/day].

* $1$ [м$^3$/сек] = $543439$ [bbl/day] = $86400$ [м$^3$/сут]
* $1$ [м$^3$/сут] = $\dfrac{1}{86400}$ [м$^3$/сек] $= 1.157407 \cdot 10^{-5}$ [м$^3$/сек]
* $1$ [bbl/day] = $0.15898$ [м$^3$/сут]

In [8]:
# выведем некоторые переводные коэффициенты для объемных расходов
print(f'Одни [сут] = {24*60*60} = {const.day}  [сек]')
print(f'Один [м3/сут] = {1/const.day} [м3/сек]')
print(f'Один баррель в день [bbl/day] = {const.bbl} [м3/сут]')
print(f'Один баррель в день [bbl/day] = {const.bbl/const.day} [м3/сек]')
print(f'Один [м3/сут] = {1/const.day} [м3/сек]')

Одни [сут] = 86400 = 86400.0  [сек]
Один [м3/сут] = 1.1574074074074073e-05 [м3/сек]
Один баррель в день [bbl/day] = 0.15898729492799998 [м3/сут]
Один баррель в день [bbl/day] = 1.8401307283333331e-06 [м3/сек]
Один [м3/сут] = 1.1574074074074073e-05 [м3/сек]


### Проницаемость $k$

В СИ измеряется в [м$^2$], в практических метрических единицах измеряется в [мД], в американских промысловых единицах измеряется в [mD].

Определение: в пористой среде с проницаемостью в один Дарси для поддержания течения жидкости с динамической вязкостью 1 сП со скоростью фильтрации 1 см/с необходимо поддерживать перепад давления жидкости приблизительно в одну атмосферу на 1 см вдоль направления течения. При использовании физической атмосферы для расчета перепада давления (физическая атмосфера = 101 325 Па) единица проницаемости равняется приблизительно 0.986923 мкм². 

В отечественной литературе при определении дарси в качестве величины атмосферы было принято использовать техническую атмосферу (1 кгс/см² = 98 066,5 Па), так что для величины дарси получалось значение приблизительно 1,02 мкм², причём эпизодические случаи использования западного определения дарси специально отмечались [[ru.wikipedia.org/wiki/Дарси](https://ru.wikipedia.org/wiki/%D0%94%D0%B0%D1%80%D1%81%D0%B8)]. Согласно  ГОСТ 26450.2-85 величины 1 Дарси $= 0.9869⋅10^{−12}$ м².

* $1$ [м$^2$] = $1.01325 \cdot 10^{15}$ [мД]
* $1$ [мД] = $0.986923 \cdot 10^{-15}$ [м$^2$]

In [9]:
print(f'Один [мД] = {1e5/const.atm * 1e-15} [м²]')

Один [мД] = 9.86923266716013e-16 [м²]




### Вязкость $\mu$

* $1$ [Па$\cdot$с] = $1000$ [сП]
* $1$ [сП] = $10^{-3}$ [Па$\cdot$с]

### Давление $p$
* $1$ [Па] = $0.0001450$ [psi] = $0.00000987$ [атм]
* $1$ [атм] = $14.6959$ [psi] = $101325$ [Па]


In [5]:
AT = 98066.5  # technical atmosphere in Pa,  техническая атмосфера в Па
print(f'Один  [psi] в [Па] = {const.psi}')
print(f'Один  [bar] в [Па] = {const.bar}')
print(f'Один  [atm] в [Па] = {const.atm}')
print(f'Один  [at] в [Па] = {AT}')
print(f'Один  [atm] в [psi] = {const.atm/const.psi}')

Один  [psi] в [Па] = 6894.757293168361
Один  [bar] в [Па] = 100000.0
Один  [atm] в [Па] = 101325.0
Один  [at] в [Па] = 98066.5
Один  [atm] в [psi] = 14.69594877551345



### Расстояние $x$

* $1$ [м] = $3.28$ [ft]



# Размерный коэффициент для формулы Дюпюи
Используя рассчитанные выше переводные коэффициенты для различных размерных величин рассчитаем переводной коэффициент в формуле Дюпюи

$$ Q = \frac{ 2 \pi kh}{ \mu B} \frac{  \left( p_i - p \right) } {\ln{\dfrac{r_e}{r_w}} +S } $$



In [23]:
# зададим переменные sympy
Q, k, h, mu, B, pres, pwf, re, rw, S, pi = sp.symbols('Q k h mu B p_res p_wf r_e r_w S pi', real=True, positive=True)

In [24]:
# определим уравнение
eq = sp.Eq(Q, 2 * pi * k * h / (mu * B) * (pres - pwf) / (sp.ln(re/rw) + S))
eq

Eq(Q, 2*h*k*pi*(p_res - p_wf)/(B*mu*(S + log(r_e/r_w))))

In [25]:
# подставим в уравнение переводные величины
eq = eq.subs(Q, 1/const.day * Q)  # дебит, [м3/сут] в [м3/сек]
f_k = 1e5/const.atm * 1e-15 
eq = eq.subs(k, f_k * k)     # проницаемость, [мД] в [м2]
eq = eq.subs(mu, 1e-3 * mu)  # вязкость, [сП] в [Па сек]
eq = eq.subs(pres, const.atm * pres) # давление [атм] в [Па]
eq = eq.subs(pwf, const.atm * pwf) # давление [атм] в [Па]
eq = eq.subs(pi, const.pi)

display(eq)


Eq(1.15740740740741e-5*Q, 6.20102176874373e-12*h*k*(101325.0*p_res - 101325.0*p_wf)/(B*mu*(S + log(r_e/r_w))))

Решим полученное уравнение относительно Q и упростим средствами `sympy`

In [26]:
eq1 = sp.simplify(sp.solve(eq,Q)[0])
display(sp.Eq(Q,eq1))

Eq(Q, 0.0542867210540316*h*k*(p_res - p_wf)/(B*mu*(S + log(r_e/r_w))))

Выделим полученную константу в явном виде и найдем обратную величину - это и будет необходимый нам переводной коэффициент.

In [27]:
f = 1/eq1.args[0]
f

18.4207110060064

По умолчанию `sympy` автоматически организует порядок элементов в своих выражениях. Этот порядок может отличаться от привычного - хотя и суть формул при этом не меняется. Применяя некоторые хитрости можно заставить `sympy` вывести выражения в приемлимом виде.

In [28]:
a = sp.symbols('a')
eq2 = eq1.subs(eq1.args[0],1/a)
with sp.evaluate(False):
    display(eq2.subs(a, f))

h*k*(p_res - p_wf)/(18.4207110060064*B*mu*(S + log(r_e/r_w)))

Но иногда результат проще переписать руками в нужном виде. В итоге уравнение Дюпюи в практических метрических единицах измерения примет вид.

$$ Q = \frac{ kh}{18.42  \mu B} \frac{  \left( p_i - p \right) } {\ln{\dfrac{r_e}{r_w}} +S } $$

где

* $Q$ - дебит скважины на поверхности, приведенный к нормальным условиям, ст. м$^3$/сут
* $\mu$ - вязкость нефти в пласте, сП
* $B$ - объемный коэффициент нефти, м$^3$/м$^3$
* $P_{res}$ - пластовое давление или давление на контуре с радиусом $r_e$, атма
* $P_{wf}$ - давление забойное, атма
* $k$ - проницаемость, мД
* $h$ - мощность пласта, м
* $r_e$ - внешний контур дренирования скважины, м
* $r_w$ - радиус скважины, м
* $S$ - скин-фактор скважины, м